docs/proposals/PROPOSAL_D.md

```diff
@@ -0,0 +1,480 @@
+# Proposal D - Namespaced and Graph-based Compatibility Metadata Schema
+
+> Also maintained by Compatibility Communities
+
+[Proposal C](https://github.com/opencontainers/wg-image-compatibility/pull/8) defines an example compatibility specification that includes namespaced metadata for a compatibility interest group. Missing from that proposal is a way to represent relationships between terms for within a namespace, which would be ideal for representing the compatibility metadata in a graph. Thus, here we want to describe the higher level compatibility definition that defines the allowed terms and relationships for a specific compatibility interest group metadata space, and use the [JSON Graph format](https://jsongraphformat.info/) (JGF) to map these attributes into a graph space. In summary, there are three things here:
+
+1. A **compatibility schema** is defined by a compatibility interest group, and  defines a known set of metadata attributes and relationships between them. There would be one copy of this, in a standard location, one source of truth (versioned).
+2. An extraction of metadata attributes for one or more compatibility schemas, and for a specific container image or application is a **compatibility spec**
+3. The schema is validated by the JSON Graph format (V2) while the spec is validated by [a custom schema that we provide](https://github.com/supercontainers/compspec/blob/main/schema.json)
+
+Note that this proposal has a matching prototype at [compspec-go](https://github.com/supercontainers/compspec.go) for which full [experiments have been run](https://github.com/rse-ops/lammps-matrix/blob/main/experiment/flux-operator/img/amd64/lammps-reasons-failure_lammps-reasons-failure.png).
+
+## Why do we want a graph?
+
+There are two use cases I'm currently considering for these compatibility metadata:
+
+- **image selection**: is the primary use case. At face value a graph might not seem valuable, because an algorithm could very simply (in a brute force manner) match a set of attributes to a desired set. And to be clear, that use case is not out of scope - the metadata attributes are provided in the compatibility spec in an easy to read and parse, flat format. However, given that we have these terms map into a graph defined by the schema, we can optimize the image selection process by pruning off entire subtrees (that are impossible, for example) before we even consider more detailed metadata. A scaled image selection task would be more computationally feasible than if using a brute force matching approach. Such an example might be to ask for higher level queries across all images in a registry.
+
+- **scheduling**: Schedulers can often represent resources as graphs (e.g., Flux Framework), and this community would want to be able to easily map one or more compatibility schemas into their scheduler graphs.
```

Comment:
> The metadata are already in hierarchical format (domain name). Describing the domain levels relations over a graph is redundant.

Not necessarily. Given the very simple relationship "contains" possibly you can get all of that from the structure. But there is no reason you can't have other relationship types. You might say requires / binds / negates or even use it as an attribute like hasAlias. Arguably that would be JGF metadata but the point is that the owning group can define whatever relationships they like for their needs.

Reply:
> Additionally, I disagree that we should adjust the schema for specific software types (in this case scheduler). As I mentioned in the past, I think we should build a schema by thinking of compatibility representation in the first place instead of influencing it by specific types of software because they could adopt it more easier than others.

When the software metadata is important for running the container on the host, period, it is arguably just as important as a driver or kernel feature. We should not be placing our biases on what is important.

> Having the above in mind, the question is, is the domain format good enough to represent schema? I think yes.

I disagree.

Reply:
Simple examples could be power or IO patterns. There would be namespaces of attributes for those. You'd want to say something like "node A draws from / supplies to node types B...C...D." Aside from those two being in separate namespaces (e.g., power describing physical resources) you likely are going to have the same kind of edge (relationship) describing more than one thing. It's not just a flat "this node A contains B" kind of thing. For IO, you'd have a graph that describes [patterns](https://scholar.google.com/citations?view_op=view_citation&hl=en&user=3B6sHUcAAAAJ&sortby=pubdate&citation_for_view=3B6sHUcAAAAJ:qUcmZB5y_30C) and you might want to say "this container is going to be heavy for this particular read pattern and should be used with these kinds of filesystems." That is also not a traditional parent/child or 1:1 relationship.

Reply:
> When the software metadata is important for running the container on the host, period, it is arguably just as important as a driver or kernel feature. We should not be placing our biases on what is important.

To value more one interest group against another one is not what I meant by the comment I made.
I meant to build a generic enough spec that can cover compatibility representation and its relations for all organizations that are interested in using it.

> (e.g., Flux Framework), and this community would want to be able to easily map one or more compatibility schemas into their scheduler graphs.

To me the above reasoning to the question "Why do we want a graph?" is biased. As I understood from the sentence,  we have to build the specification in a certain way because organization A would want to be able to easily map compatibility schemas into their schedeuler graphs (in this case Flux Framework), that sounds biased to me. For instance, in my case, I am also interested in a node reconfiguration but I don't want to adjust the spec that is easier to adopt by OpenShift's cluster node tuning operator or machine config operator.

That's how I read that. Maybe I misunderstood what you tried to say in there.

Anyway, the following example is valid to me:

> Simple examples could be power or IO [patterns](https://scholar.google.com/citations?view_op=view_citation&hl=en&user=3B6sHUcAAAAJ&sortby=pubdate&citation_for_view=3B6sHUcAAAAJ:qUcmZB5y_30C). There would be namespaces of attributes for those. You'd want to say something like "node A draws from / supplies to node types B...C...D." Aside from those two being in separate namespaces (e.g., power describing physical resources) you likely are going to have the same kind of edge (relationship) describing more than one thing. It's not just a flat "this node A contains B" kind of thing. For IO, you'd have a graph that describes patterns and you might want to say "this container is going to be heavy for this particular read pattern and should be used with these kinds of filesystems." That is also not a traditional parent/child or 1:1 relationship.

In our both proposals we expressed different requirements about where to build graph representation. I proposed on the spec level that is attached to the artifact. You on the schema level. I think that's a good point for the spec/schema discussion.